### DISCUSSION TITLE
2fa support

### DISCUSSION THREAD

**Discussion by gaurishhs** [go-pkgz/auth#148] [Q&A] [answered]
Hi! Is it possible to use Multifactor authentication (TOTP, SMS) with this package. Wonderful library!

**Comment by umputun** [comment, marked as answer]
moved to discussions

**Reply by gaurishhs** [reply]
Hi! Is it possible for this library to support multi factor auth? Thanks for the promptness

**Reply by umputun** [reply]
I think OTP on the provider side has nothing to do with this library. I have not tried it personally, but I expect any provider supporting OTP to work transparently with this auth and handles multifactor before it hits the callback function.

**Reply by gaurishhs** [reply]
Okay.. Does the library store users sessions? If yes how so? Are they stored in the memory?

**Reply by umputun** [reply]
The library is not for session-based auth but rather JWT, which is "stored" in the secured cookie

**Reply by gaurishhs** [reply]
So basically no memory storage.. thanks a lot.. Just one last question, Can you provide me an example avatar store? i can understand what do i have to return in the `store.PUT` call and other calls

**Reply by umputun** [reply]
Several implementations of the avatar store have already been provided; see https://github.com/go-pkgz/auth/blob/master/avatar

To make your custom avatar store, you should implement the interface https://github.com/go-pkgz/auth/blob/master/avatar/store.go#L31

And this is how you use/pass it in - https://github.com/go-pkgz/auth/blob/master/_example/main.go#L49

**Reply by gaurishhs** [reply]
Alright thanks

**Reply by gaurishhs** [reply]
@umputun Can you tell me what is `aud` in direct provider? And should i encrypted my password before sending it to the server? 

Also, If i want to have email verification with direct method how would i go about it?

**Reply by gaurishhs** [reply]
I think its for support multiple audiences, But since i don't need multiple audiences what should i do?

**Reply by umputun** [reply]
pls, check the example I have linked above and also see readme

**Comment by gaurishhs** [comment]
Ok.. Thanks a lot for your help. Appreciate it
